src/prefect/agent/docker/agent.py

```diff
@@ -67,6 +68,23 @@ def __init__(
             )
             raise exc
 
+    def heartbeat(self) -> None:
+        try:
+            if not self.docker_client.ping():
+                raise RuntimeError("Unexpected Docker ping result")
+            if self.failed_connections > 0:
+                self.logger.info("Reconnected to Docker daemon")
+            self.failed_connections = 0
+        except Exception as exc:
+            self.logger.warning("Failed heartbeat: {}".format(repr(exc)))
+            self.failed_connections += 1
+
+        if self.failed_connections >= 3:
```

Comment:
I think we should bump this number up. At an iteration of 10s a daemon restart will take probably around a minute to occur. I vote for increasing this to 6 for a full minute of retries

Reply:
Actually the more I think of it the more inclined I am to say we should exit at all 🤔 I think it should continue to make connection until a daemon is found since there is no external process responsible for restarting this it could go down on a long restart of the daemon. Thoughts?

Reply:
I feel that the agent should not be pulling work if there is risk that the work will never be done. Making an attempt to reconnect makes sense at the least --we want to be somewhat resilient to disconnects and restarts.

So it ends up that a daemon restart doesn't take that much time. On a linux system where there is not an underlying VM it should be much less than 30 seconds. On Windows/OSX, this restart time includes the time it takes to restart the VM. Initial tests on OSX shows that it squeaks in under 30 seconds... I see no problem in increasing the number of reconnects to allow for more time here, and a minute seems reasonable.

The one issue with the approach on this PR is that it is somewhat abusing the heartbeat mechanism. Agent heartbeating is intended to indicate to external orchestrators or process managers that the agent is still alive. The way that heartbeating is being used in this PR is to measure liveliness of downstream dependences (in this case the Docker daemon) and take action appropriately (logging and exiting, essentially). (in short: I realize that this is overloading the heartbeat for other purposes)

Reply:
Yea, with the new agent management screens I'd rather the agent exit so that users can see "this agent is no longer capable of submitting work"

Reply:
Okay that all makes sense to me! Let's bump to a minute 👍